### PR TITLE
Enable RMM plugin

### PR DESCRIPTION
**Pull request by hcho3**
Supersedes #1 (just handling RMM plugin)

This PR makes minimal changes to the current recipe to enable the RMM plugin.

### CODE REVIEW

**Comment by jakirkham** [comment]
So that we can get `librmm` from the RAPIDS channels, would suggest we add this to [`conda_build_config.yaml`]( https://github.com/rapidsai/xgboost-feedstock/blob/d7c47c83ed219ad5396d79bdd33fea2058a06110/recipe/conda_build_config.yaml ) above `channel_targets`

```yaml
channel_sources:
  - rapidsai,rapidsai-nightly,conda-forge,nvidia
```

Then would [re-render]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks )

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 {% set name = "xgboost" %}
 {% set version = "1.7.4" %}
+{% set build_number = "0" %}
+{% set rapids_version = "23.06.00" %}
```
Comment (jakirkham) [inline recipe/meta.yaml:4]: Should we update this for the upcoming release?

```suggestion
{% set rapids_version = "23.08.00" %}
```

**Comment by hcho3** [comment]
I'm getting error `nccl-2.18.3.1-h3a97aeb_0 is excluded by strict repo priority`. What can we do fix it?

**Comment by jakirkham** [comment]
Think we want to add

```
channel_priority: flexible
```

Here

https://github.com/rapidsai/xgboost-feedstock/blob/d7c47c83ed219ad5396d79bdd33fea2058a06110/.scripts/build_steps.sh#L24-L32

File: recipe/meta.yaml [hunk 2]
```diff
@@ -32,8 +36,13 @@ requirements:
     - ninja
   host:
     - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-    - nccl       # [cuda_compiler_version != "None"]
+    - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:39]:
```suggestion
    - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -32,8 +36,13 @@ requirements:
...
+    # Require NCCL 2.18+ to support CUDA 12
+    - nccl >=2.18.1  # [cuda_compiler_version != "None"]
     - llvm-openmp  # [osx]
+    - librmm ={{ rapids_version }}  # [linux]
+  run:
+    - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:45]: This is handled [via `run_exports` from `cuda-cudart-dev`]( https://github.com/conda-forge/cuda-cudart-feedstock/blob/67400f12d49dae4b83b5a0c44751e53c27e9c7c0/recipe/meta.yaml#L101-L103 )

```suggestion
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -55,10 +64,15 @@ outputs:
       host:
         - llvm-openmp  # [osx]
         - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-        - nccl       # [cuda_compiler_version != "None"]
+        - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:67]:
```suggestion
        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
```

File: recipe/meta.yaml [hunk 5]
```diff
@@ -55,10 +64,15 @@ outputs:
...
+        # Require NCCL 2.18+ to support CUDA 12
+        - nccl >=2.18.1  # [cuda_compiler_version != "None"]
+        - librmm ={{ rapids_version }}  # [linux]
       run:
         - llvm-openmp  # [osx]
+        - librmm ={{ rapids_version }}  # [linux]
         - __cuda  # [cuda_compiler_version != "None"]
+        - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:75]:
```suggestion
```

File: recipe/meta.yaml [hunk 6]
```diff
@@ -129,11 +143,14 @@ outputs:
         # install llvm-openmp in the top-level host env above
         - llvm-openmp  # [osx]
         - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-        - nccl       # [cuda_compiler_version != "None"]
+        - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:146]:
```suggestion
        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
```

File: recipe/meta.yaml [hunk 7]
```diff
@@ -129,11 +143,14 @@ outputs:
...
+        # Require NCCL 2.18+ to support CUDA 12
+        - nccl >=2.18.1  # [cuda_compiler_version != "None"]
       run:
         - python
         - {{ pin_subpackage('py-xgboost', exact=True) }}
         - __cuda  # [cuda_compiler_version != "None"]
+        - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:153]:
```suggestion
```

File: recipe/meta.yaml [hunk 8]
```diff
@@ -179,7 +196,9 @@ outputs:
         - r-jsonlite
         - r-knitr
         - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-        - nccl       # [cuda_compiler_version != "None"]
+        - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:199]:
```suggestion
        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
```

File: recipe/meta.yaml [hunk 9]
```diff
@@ -190,6 +209,7 @@ outputs:
         - r-magrittr
         - r-jsonlite
         - __cuda  # [cuda_compiler_version != "None"]
+        - cuda-cudart      # [(cuda_compiler_version or "").startswith("12")]
```
Comment (jakirkham) [inline recipe/meta.yaml:212]:
```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Thanks Philip! 🙏

Regarding `cudart`, would use `cuda-cudart-dev` in `host` and then drop `cuda-cudart` from `run` (as it is handled [via `run_exports` from `cuda-cudart-dev`]( https://github.com/conda-forge/cuda-cudart-feedstock/blob/67400f12d49dae4b83b5a0c44751e53c27e9c7c0/recipe/meta.yaml#L101-L103 ))

File: recipe/conda_build_config.yaml [hunk 10]
```diff
@@ -1,2 +1,4 @@
+channel_sources:
```
Comment (jakirkham) [inline recipe/conda_build_config.yaml:1]:
```suggestion
# First CUDA 12 build of NCCL was 2.17.1.1
# xref: https://github.com/conda-forge/nccl-feedstock/pull/85
nccl:
  - 2.17.1.1

channel_sources:
```

File: recipe/meta.yaml [hunk 11]
```diff
@@ -32,8 +36,11 @@ requirements:
     - ninja
   host:
     - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-    - nccl       # [cuda_compiler_version != "None"]
+    - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
+    # Require NCCL 2.18+ to support CUDA 12
+    - nccl >=2.18.1  # [cuda_compiler_version != "None"]
```
Comment (jakirkham) [inline recipe/meta.yaml:41]:
```suggestion
    - nccl             # [cuda_compiler_version != "None"]
```

File: recipe/meta.yaml [hunk 12]
```diff
@@ -55,9 +62,13 @@ outputs:
       host:
         - llvm-openmp  # [osx]
         - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-        - nccl       # [cuda_compiler_version != "None"]
+        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
+        # Require NCCL 2.18+ to support CUDA 12
+        - nccl >=2.18.1  # [cuda_compiler_version != "None"]
```
Comment (jakirkham) [inline recipe/meta.yaml:67]:
```suggestion
        - nccl             # [cuda_compiler_version != "None"]
```

File: recipe/meta.yaml [hunk 13]
```diff
@@ -179,7 +192,9 @@ outputs:
         - r-jsonlite
         - r-knitr
         - cudatoolkit      # [(cuda_compiler_version or "").startswith("11")]
-        - nccl       # [cuda_compiler_version != "None"]
+        - cuda-cudart-dev  # [(cuda_compiler_version or "").startswith("12")]
+        # Require NCCL 2.18+ to support CUDA 12
+        - nccl >=2.18.1  # [cuda_compiler_version != "None"]
```
Comment (jakirkham) [inline recipe/meta.yaml:197]:
```suggestion
        - nccl             # [cuda_compiler_version != "None"]
```

**Review by jakirkham — COMMENTED** [review]
Thanks Philip! 🙏

Think we can also state the `nccl` version constraint once in `conda_build_config.yaml` and then reuse it throughout. Should minimize the number of places we need to track it

Note: The first CUDA 12 build of NCCL was 2.17.1.1. Hence the version was rewound slightly

File: recipe/conda_build_config.yaml [hunk 14]
```diff
@@ -1,2 +1,9 @@
+# First CUDA 12 build of NCCL was 2.17.1.1
+# xref: https://github.com/conda-forge/nccl-feedstock/pull/85
+nccl:
+  - 2.17.1.1
+
+channel_sources:
+  - rapidsai,rapidsai-nightly,conda-forge,nvidia
```
Comment (jakirkham) [inline recipe/conda_build_config.yaml:7]:
```suggestion
  - rapidsai,rapidsai-nightly,conda-forge
```

File: recipe/conda_build_config.yaml [hunk 15]
```diff
@@ -1,2 +1,9 @@
+# First CUDA 12 build of NCCL was 2.17.1.1
+# xref: https://github.com/conda-forge/nccl-feedstock/pull/85
+nccl:
+  - 2.17.1.1
+
```
Comment (jakirkham) [inline recipe/conda_build_config.yaml:5]:
```suggestion
```

**Review by jakirkham — COMMENTED** [review]
Thanks Philip! 🙏

It looks like the `nvidia` channel is being used for the CTK packages, which causes us a problem in the tests since the CTK packages in the `nvidia` channel don't have all the fixes that were added to `conda-forge` CTK packages

Also suspect this is why there were issues getting only `nccl` packages before

Looking through CI it appears nothing else is gotten from the `nvidia` channel. So suggesting we drop it

Then will need another re-render

This should fix [the CI issue]( https://github.com/rapidsai/xgboost-feedstock/actions/runs/5580000563/jobs/10196312663?pr=5#step:4:4864 )

**Comment by jakirkham** [comment]
Looks like CI failed due strict channel priority. So we need to readd [this change]( https://github.com/rapidsai/xgboost-feedstock/pull/5#issuecomment-1635130310 )

**Review by jakirkham — APPROVED** [review]
Thanks Philip! 🙏

Think this is ready to go. Let's follow up on anything else in a subsequent PR 🙂

**Pull request merged**